src/DiffBuilder/Diff.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Phpcq\RepositoryBuilder\DiffBuilder;
+
+use LogicException;
+use Phpcq\RepositoryBuilder\Repository\Tool;
+
+final class Diff implements DiffInterface
+{
+    /**
+     * @var DiffInterface[]|null
+     */
+    private ?array $differences;
+
+    /**
+     * @param Tool[]|null $old Tools indexed by tool name.
+     * @param Tool[]|null $new Tools indexed by tool name.
+     *
+     * @return Diff|null
+     */
+    public static function diff(?array $old, ?array $new): ?Diff
+    {
+        if (null === $old && null === $new) {
+            throw new LogicException('new value and old value must not both be null.');
+        }
+
+        // New repository, add all tools as new.
+        if (null === $old) {
+            $changes = [];
+            foreach ($new as $tool) {
+                $changes[$tool->getName()] = ToolAddedDiff::diff($tool);
+            }
+            return new Diff($changes);
```

Comment:
What if `$new` is an empty array? An empty diff would be created. `deepCompare` would return null in such a case. Both methods should act similar in the case of an empty diff.

Reply:
I have no idea how this might happen in real world scenarios but implemented in 691ceb4f15fd738b6593bb23b2338d74c5dc5ebb.